services/s3/src/main/java/software/amazon/awssdk/services/s3/S3Utilities.java

```diff
@@ -0,0 +1,231 @@
+                                                   .httpMethod(SdkHttpMethod.HEAD)
+                                                   .build();
+
+        SdkHttpFullRequest.Builder builder = ProtocolUtils.createSdkHttpRequest(operationInfo, endpoint);
+
+        // encode bucket
+        builder.encodedPath(PathMarshaller.NON_GREEDY.marshall(builder.encodedPath(),
+                                                               "Bucket",
+                                                               getUrlRequest.bucket()));
+
+        // encode key
+        builder.encodedPath(PathMarshaller.GREEDY.marshall(builder.encodedPath(), "Key", getUrlRequest.key()));
+
+        return builder.build();
+    }
+
+    /**
+     * Apply the {@link S3Configuration} options on the given #sdkHttpFullRequest.
+     * Uses {@link EndpointAddressInterceptor} class instead of duplicating the same functionality here.
+     */
+    private SdkHttpRequest applyEndpointInterceptor(SdkHttpFullRequest sdkHttpFullRequest,
+                                                    GetUrlRequest getUrlRequest,
+                                                    Region resolvedRegion) {
+        ExecutionAttributes executionAttributes = new ExecutionAttributes();
+        executionAttributes.putAttribute(AwsSignerExecutionAttribute.SERVICE_CONFIG, s3Configuration);
+        executionAttributes.putAttribute(AwsExecutionAttribute.AWS_REGION, resolvedRegion);
+
+        EndpointAddressInterceptor interceptor = new EndpointAddressInterceptor();
+        return interceptor.modifyHttpRequest(constructModifyHttpRequest(sdkHttpFullRequest, getUrlRequest),
+                                             executionAttributes);
+    }
+
+    // Used as input for EndpointAddressInterceptor#modifyHttpRequest method
+    private static Context.ModifyHttpRequest constructModifyHttpRequest(SdkHttpFullRequest httpFullRequest,
+                                                                        GetUrlRequest getUrlRequest) {
+        return new Context.ModifyHttpRequest() {
+            @Override
+            public SdkHttpRequest httpRequest() {
+                return httpFullRequest;
+            }
+
+            @Override
+            public Optional<RequestBody> requestBody() {
+                return Optional.empty();
+            }
+
+            @Override
+            public Optional<AsyncRequestBody> asyncRequestBody() {
+                return Optional.empty();
+            }
+
+            @Override
+            public SdkRequest request() {
+                return getUrlRequest;
+            }
+        };
+    }
+
+    /**
+     * Builder class to construct {@link S3Utilities} object
+     */
+    public static final class Builder {
+        private Region region;
+
+        private S3Configuration s3Configuration;
+
+        private Builder() {
+        }
+
+        /**
+         * Configure the region with which the SDK should communicate.
```

Comment:
This comment probably needs to be updated. The SDK won't be doing any communication.

Reply:
Done